File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/adapter/TrackDetailAdapter.java
package com.andruid.magic.discodruid.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.andruid.magic.discodruid.R;
import com.andruid.magic.discodruid.model.Track;
import com.andruid.magic.discodruid.viewholder.TrackDetailViewHolder;
import com.thesurix.gesturerecycler.GestureAdapter;

public class TrackDetailAdapter extends GestureAdapter<Track,TrackDetailViewHolder> {
    private Context mContext;

    public TrackDetailAdapter(Context mContext) {
        this.mContext = mContext;
    }

    @NonNull
    @Override
    public TrackDetailViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, final int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.layout_track_detail,viewGroup,false);
        return new TrackDetailViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull TrackDetailViewHolder trackDetailViewHolder, int pos) {
        Track track = getItem(pos);
        trackDetailViewHolder.setItem(track,mContext);
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/ui/adapter/TrackDetailAdapter.kt
package com.andruid.magic.discodruid.ui.adapter

import android.content.Context
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.andruid.magic.discodruid.ui.viewholder.TrackDetailViewHolder
import com.andruid.magic.medialoader.model.Track
import kotlinx.coroutines.CoroutineScope

private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Track>() {
    override fun areItemsTheSame(oldItem: Track, newItem: Track) =
        oldItem.audioId == newItem.audioId

    override fun areContentsTheSame(oldItem: Track, newItem: Track) =
        oldItem == newItem
}

class TrackDetailAdapter(
    private val context: Context,
    private val scope: CoroutineScope
) : ListAdapter<Track, TrackDetailViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        TrackDetailViewHolder.from(parent)

    override fun onBindViewHolder(holder: TrackDetailViewHolder, position: Int) {
        getItem(position)?.let { track ->
            holder.bind(context, scope, track)
        }
    }
}